protocols/noise/src/util.rs

```diff
@@ -0,0 +1,128 @@
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{keys::{Curve25519, Keypair, SecretKey, PublicKey}, NoiseError};
+use rand::FromEntropy;
+
+pub(crate) fn to_array(bytes: &[u8]) -> Result<[u8; 32], NoiseError> {
+    if bytes.len() != 32 {
+        return Err(NoiseError::InvalidKey)
+    }
+    let mut m = [0; 32];
+    m.copy_from_slice(bytes);
+    Ok(m)
+}
+
+/// Custom `snow::CryptoResolver` which uses `ring` as much as possible.
+pub(crate) struct Resolver;
+
+impl snow::resolvers::CryptoResolver for Resolver {
+    fn resolve_rng(&self) -> Option<Box<dyn snow::types::Random>> {
+        Some(Box::new(Rng(rand::rngs::StdRng::from_entropy())))
+    }
+
+    fn resolve_dh(&self, choice: &snow::params::DHChoice) -> Option<Box<dyn snow::types::Dh>> {
+        if let snow::params::DHChoice::Curve25519 = choice {
+            Some(Box::new(X25519 { keypair: Keypair::gen_curve25519() }))
+        } else {
+            None
+        }
+    }
+
+    fn resolve_hash(&self, choice: &snow::params::HashChoice) -> Option<Box<dyn snow::types::Hash>> {
+        snow::resolvers::RingResolver.resolve_hash(choice)
+    }
+
+    fn resolve_cipher(&self, choice: &snow::params::CipherChoice) -> Option<Box<dyn snow::types::Cipher>> {
+        snow::resolvers::RingResolver.resolve_cipher(choice)
+    }
+}
+
+/// Wrapper around a CPRNG to implement `snow::Random` trait for.
+struct Rng(rand::rngs::StdRng);
+
+impl rand::RngCore for Rng {
+    fn next_u32(&mut self) -> u32 {
+        self.0.next_u32()
+    }
+
+    fn next_u64(&mut self) -> u64 {
+        self.0.next_u64()
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
+impl rand::CryptoRng for Rng {}
+
+impl snow::types::Random for Rng {}
+
+/// X25519 Diffie-Hellman key agreement.
+struct X25519 {
+    keypair: Keypair<Curve25519>
+}
+
+impl snow::types::Dh for X25519 {
+    fn name(&self) -> &'static str { "25519" }
+
+    fn pub_len(&self) -> usize { 32 }
+
+    fn priv_len(&self) -> usize { 32 }
+
+    fn pubkey(&self) -> &[u8] {
+        &self.keypair.public().as_ref()
+    }
+
+    fn privkey(&self) -> &[u8] {
+        &self.keypair.secret().as_ref()
+    }
+
+    fn set(&mut self, privkey: &[u8]) {
+        let mut s = [0; 32];
+        s.copy_from_slice(privkey);
+        let secret = SecretKey::new(s);
+        let public = secret.public();
+        self.keypair = Keypair::new(secret, public)
+    }
+
+    fn generate(&mut self, rng: &mut snow::types::Random) {
+        let mut s = [0; 32];
+        rng.fill_bytes(&mut s);
+        let secret = SecretKey::new(s);
+        let public = secret.public();
+        self.keypair = Keypair::new(secret, public)
+    }
+
+    fn dh(&self, pubkey: &[u8], out: &mut [u8]) -> Result<(), ()> {
+        let mut p = [0; 32];
+        p.copy_from_slice(&pubkey[.. 32]);
```

Comment:
Here you do `&pubkey[..32]` while in `set` you do just `privkey`. I'd prefer to add a verification that the lengths of `pubkey` and `out` are `32`.

Reply:
The `pubkey` that is passed in is longer than 32 bytes, hence the need to restrict it. When called from [`HandshakeState::dh`](https://github.com/mcginty/snow/blob/953deb8f09c4f0908d4c820f61e9a4a84c6ac1ae/src/handshakestate.rs#L136) it uses its `re` or `rs` slices which are 56 bytes each.